nncf/torch/sparsity/movement/structured_mask_handler.py

```diff
@@ -0,0 +1,211 @@
+        temp_shape = list(itertools.chain(*zip(list(structured_mask_shape), list(grain_size))))
+        structured_mask = self.sparsifier_operand.weight_ctx.binary_mask.detach().clone()
+        structured_mask = structured_mask.reshape(temp_shape)
+        structured_mask = structured_mask.amax(dim=(tuple((np.arange(len(self.sparsifier_operand.weight_ctx.binary_mask.shape)) * 2 + 1))))
+        if self.sparsifier_operand.prune_bias is True:
+            structured_bias_mask_shape = structured_mask_shape[0]
+            structured_bias_mask = self.sparsifier_operand.bias_ctx.binary_mask.detach().clone()
+            structured_bias_mask = structured_bias_mask.reshape((structured_bias_mask_shape, -1))
+            structured_bias_mask = structured_bias_mask.amax(dim=1)
+            # dim_aligned = structured_bias_mask.repeat(structured_mask.shape[1]).reshape(-1, structured_mask.shape[1])
+            # structured_mask = structured_mask.logical_or(dim_aligned).to(torch.float32)
+            prunable_rows = structured_bias_mask.logical_or(structured_mask.amax(dim=1))  # preserve a row when either bias mask is 1 or weight mask row amax is 1
+            prunable_cols = structured_mask.amax(dim=0)
+            structured_mask = prunable_rows.unsqueeze(1) * prunable_cols.unsqueeze(0)
+        self.independent_structured_mask = structured_mask
+        return structured_mask
+
+    def _inflate_structured_mask(self, structured_mask: torch.Tensor, grid_size: Tuple[int, int]) -> torch.Tensor:
+        assert len(structured_mask.shape) == len(grid_size), "Unmatching dimension"
+        inflated_mask = structured_mask.clone()
+        for axis, repeat_times in enumerate(grid_size):
+            inflated_mask = inflated_mask.repeat_interleave(repeat_times, dim=axis)
+        return inflated_mask
+
+    def populate_dependent_structured_mask_to_operand(self):
+        structured_mask_inflated = self._inflate_structured_mask(self.dependent_structured_mask, self.grid_size)
+        self.sparsifier_operand.weight_ctx.binary_mask = structured_mask_inflated
+        if self.sparsifier_operand.prune_bias is True:
+            self.sparsifier_operand.bias_ctx.binary_mask = structured_mask_inflated.amax(dim=1)
+
+
+class StructuredMaskHandler:
+
+    def __init__(self,
+                 sparsified_module_info_groups: List[SparsifiedModuleInfoGroup],
+                 strategy):
+        self.sparsified_module_info_groups = sparsified_module_info_groups
+        self.strategy = strategy
+        self.strategy_by_group_type = strategy.strategy_by_group_type
+        self._structured_mask_ctx_by_group_type = self._create_structured_mask_ctx_by_group_type()
+
+    def _get_group_of_prunable_sparsified_module_info(self) -> List[SparsifiedModuleInfoGroup]:
+        module_2_sparse_module_info_map = {sparse_info.module: sparse_info for sparse_info in self.sparsified_module_info}
+        building_blocks, _ = get_building_blocks(self.model,
```

Comment:
It would be great to have a `get_building_blocks` call in a single place.
Currently, it's called in 4 places. 

https://github.com/openvinotoolkit/nncf/pull/1319/files#diff-e7e7decd8a9de2e00ac17eace8c941f15bb9d2b7676d077e4e8d5f7d24d3222dR330
https://github.com/openvinotoolkit/nncf/pull/1319/files#diff-e7e7decd8a9de2e00ac17eace8c941f15bb9d2b7676d077e4e8d5f7d24d3222dR355
https://github.com/openvinotoolkit/nncf/pull/1319/files#diff-e7e7decd8a9de2e00ac17eace8c941f15bb9d2b7676d077e4e8d5f7d24d3222dR415
https://github.com/openvinotoolkit/nncf/pull/1319/files#diff-efe63e2fdf2856e83f319f649f138cd90deb682e4c29e9a4bd5b14778acc95aeR142

Reply:
BTW, any reason why it's needed in multiple places? @vuiseng9 @yujiepan-work

Reply:
@ljaljushkin it will only be called once. The other two callers are artifacts for debug during our development, will refactor. @yujiepan-work we miss to capture this in the list. Thanks Nikolay for pointing this out! :)

Reply:
the other duplication was due to two `_get_group_of_prunable_sparsified_module_info` in separate class, @yujiepan-work let's consolidate to one.

Reply:
@ljaljushkin there is only one instance now `get_building_blocks` after refactoring by Yujie.

Reply:
👍